src/compiler/checker.ts

```diff
@@ -17937,6 +17937,16 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (contains(types, wildcardType)) {
             return wildcardType;
         }
+        if (
+            texts.length === 2 && texts[0] === "" && texts[1] === ""
+            // literals (including string enums) are stringified below
+            && !(types[0].flags & TypeFlags.Literal)
+            // infer T extends StringLike can't be unwrapped eagerly

```

Comment:
Do you have an example of when this matters? I.e. an example of why we shouldn't unwrapp `infer T`?

Reply:
it's mainly because of how string enums might be "stringified" by template literal types
```ts
const enum StringLiteralEnum { Zero = "0", True = "true", False = "false", Undefined = "undefined", Null = "null" }

type TStringLiteralEnum0 = "0" extends `${infer T extends StringLiteralEnum}` ? T : never;
//   ^? type TStringLiteralEnum0 = StringLiteralEnum.Zero
type TStringLiteralEnum0Unwrapped = "0" extends infer T extends StringLiteralEnum ? T : never;
//   ^? type TStringLiteralEnum0Unwrapped = never
```